rubyforgood/human-essentials
# Distribution email configuration

Resolves #2340

### Description
Adds a new service called `TextInterpolatorService` that can be reused in the future in case more parts of the application need to be dynamic - specially emails.

I opted to create a service instead of using `I18n.interpolate` since `I18n` raises an exception by default when interpolating a variable that doesn't exist. This service is safer and is tested.

I still need to do the following and take decisions:

- [x] Default email content is hardcoded in the email template. Do we want to keep this?
- [x] Not sure where to put the `distribution.comment`, should this be interpolated too?
- [x] Add help text in the organization form (see below).

### Type of change

* Breaking change (fix or feature that would cause existing functionality to not work as expected)

### How Has This Been Tested?

1. Comment out line 9 of `app/mailers/distribution_mailer.rb` in order to skip validations
2. Visit `http://localhost:3000/rails/mailers/distribution_mailer/partner_mailer` to see the preview of the email
3. Visit `http://localhost:3000/diaper_bank/manage/edit` to see the form to update the default email text
4. Try using the following content to test out the feature:
```
Hiya %{partner_name}

The delivery method is: %{delivery_method}, and the following information is important:

- %{distribution_date}

This is an interpolation mistake and it will not raise an error: %{mistake}

Regards, %{partner_name}
```
5. Once updated, you should see the variables being interpolated.
6. You can try adding bold text, cursive, a list, etc.

To backfill the organizations with the default email, run the rake task `organizations:backfill:default_email_text`.

### Screenshots
![image](https://user-images.githubusercontent.com/1270156/136675580-7bf17192-0665-488c-93c5-4c2a5e0dd276.png)
![image](https://user-images.githubusercontent.com/1270156/136675585-edccebdd-a5b4-4e89-af46-018ac37cb90a.png)
